bin/identify_dormant_github_users_v2.py

```diff
@@ -0,0 +1,139 @@
+    "moj-observability-platform-bot",
+    "moj-operations-engineering-bot",
+    "mojanalytics",
+    "mojplatformsdeploy",
+    "operations-engineering-servicenow",
+    "opg-integrations",
+    "opg-use-an-lpa",
+    "opg-weblate",
+    "slack-moj",
+]
+
+
+@dataclass
+class DormantUser:
+    name: str
+    email: str | None
+
+
+def get_inactive_users_from_data_lake_ignoring_bots_and_collaborators(github_service, bot_list: list) -> list:
+    all_users = github_service.get_all_enterprise_members()
+
+    cloudtrail_service = CloudtrailService()
+    active_users = cloudtrail_service.get_active_users_for_dormant_users_process()
+
+    return [user for user in all_users if user not in list(set(active_users).union(bot_list))]
+
+
+def get_inactive_committers(gh_orgs, inactive_users_from_audit_log):
+
+    active_committers = []
+
+    for gh in gh_orgs:
+        repos_and_current_contributors = gh.get_current_contributors_for_active_repos()
+        since_datetime = (datetime.now() - timedelta(days=90))
+
+        for username in inactive_users_from_audit_log:
+            commits = False
+            if username not in active_committers:
+                commits = gh.user_has_commmits_since(
+                    login=username.lower(),
+                    repos_and_contributors=repos_and_current_contributors,
+                    since_datetime=since_datetime
+                )
+            if commits and username not in active_committers:
+                active_committers.append(username)
+
+    return list(set(inactive_users_from_audit_log).difference(set(active_committers)))
+
+
+def get_active_users_from_auth0_log_group() -> list:
+    """Operations Engineering, Cloud, Mod and Analytical Platforms all have thir own
+    Auth0 tenants. All logs are collected in a single Cloudwatch log group.
+    This function will parse the logs and return a list of active users using
+    an insights query.
+    """
+    cloudwatch_log_group = "/aws/events/LogsFromOperationsEngineeringAuth0"
+
+    cloudwatch_service = CloudWatchService(cloudwatch_log_group)
+    return cloudwatch_service.get_all_auth0_users_that_appear_in_tenants()
+
+
+def filter_out_active_auth0_users(dormant_users_according_to_github: list) -> list:
+    active_email_addresses = get_active_users_from_auth0_log_group()
+
+    dormant_users_not_in_auth0 = [
+        user.name
+        for user in dormant_users_according_to_github
+        if user.email and user.email.lower() not in active_email_addresses
+    ]
+
+    return dormant_users_not_in_auth0
+
+
+def map_usernames_to_emails(users, moj_github_org: GithubService, ap_github_org: GithubService) -> list[DormantUser]:
+    dormant_users = [
+        DormantUser(
+            user,
+            moj_github_org.get_user_org_email_address(user)
+            or ap_github_org.get_user_org_email_address(user),
+        )
+        for user in users
+    ]
+
+    return dormant_users
+
+
+def identify_dormant_github_users():
```

Comment:
I really like the readability of these function names 🚀 ❤️ 

Couple of suggestions for the main function:
- Use explicit names for each service created i.e. `MoJGitHubService = GithubService(env.get("GH_ADMIN_TOKEN"), MINISTRY_OF_JUSTICE)`
- Create all services at the top level instead of later in the process - makes it really clear what this script talks to i.e. GitHub, ClouWatch etc., and also helps the script to fail fast in case any config issues. It also means all dependencies are be injected which makes testing them easier 
- Prefer to use `main()` as the entry point function, helps to identify the entry point without the need to check which function gets called at the bottom 🙈